src/backend/distributed/connection/connection_management.c

```diff
@@ -1095,21 +1080,67 @@ AfterXactHostConnectionHandling(ConnectionHashEntry *entry, bool isCommit)
 		}
 		else
 		{
-			/* reset per-transaction state */
-			ResetRemoteTransaction(connection);
-			ResetShardPlacementAssociation(connection);
-
-			/* reset copy state */
-			connection->copyBytesWrittenSinceLastFlush = 0;
-
-			UnclaimConnection(connection);
+			/*
+			 * Preserve healthy session lifespan connections.
+			 */
+			PreserveConnection(connection);
 
 			cachedConnectionCount++;
 		}
 	}
 }
 
 
+/*
+ * ShouldShutdownConnection returns true if either one of the followings is true:
+ * - The connection is citus initiated.
+ * - Current cached connections is already at MaxCachedConnectionPerWorker
+ * - Connection is forced to close at the end of transaction
+ * - Connection is not in OK state
+ * - Connection is not idle.
```

Comment:
Connection is not idle -> A transaction is still in progress (usually because we are cancelling a distributed transaction)